syntax = "proto3";

package descarteslabs.workflows;
import "descarteslabs/common/proto/typespec/typespec.proto";
import "descarteslabs/common/proto/visualization/visualization.proto";
import "descarteslabs/common/proto/widgets/widgets.proto";

service WorkflowAPI {
  rpc UpsertWorkflow(UpsertWorkflowRequest) returns (Workflow) {}
  rpc GetWorkflow(GetWorkflowRequest) returns (Workflow) {}
  rpc GetVersion(GetVersionRequest) returns (VersionedGraft) {}
  rpc SearchWorkflows(SearchWorkflowsRequest) returns (stream Workflow) {}
  rpc DeleteWorkflow(DeleteWorkflowRequest) returns (Empty) {}
  rpc GetWmtsUrlTemplate(Empty) returns (WmtsUrlTemplateResponse) {}
}

message Workflow {
  // "email@example.com:name" - email + ":" + name
  string id = 1;

  // milliseconds since the UNIX epoch (UTC)
  int64 created_timestamp = 2;
  // milliseconds since the UNIX epoch (UTC)
  int64 updated_timestamp = 3;

  // obsolete: bool public = 5;

  string title = 9;
  string description = 8;

  repeated VersionedGraft versioned_grafts = 26;
  map<string, string> labels = 27;
  repeated string tags = 29;

  // RO
  string user = 23;
  string org = 24;
  string email = 25;
  string name = 28;
  // Template string for WMTS service endpoint, like
  // "https://workflows.descarteslabs.com/v0-5/workflow/test@example.com:foo?tile_matrix_set={TileMatrixSet}".
  // Compatible with Leaflet, OpenLayers, Python format strings.
  // Read-only: any value is ignored in `UpsertWorkflowRequest`.
  string wmts_url_template = 30;
}

message UpsertWorkflowRequest {
  string id = 1;

  // obsolete: bool public = 2;
  string title = 3;
  string description = 4;

  repeated VersionedGraft versioned_grafts = 26;
  map<string, string> labels = 27;
  repeated string tags = 29;

  bool dry_run = 50;
}

message GetWorkflowRequest { string id = 1; }

message GetVersionRequest {
  string id = 1;
  string version = 2;
}

message DeleteWorkflowRequest { string id = 1; }

message SearchWorkflowsRequest {
  string email = 1;
  string name_prefix = 2;

  // must have any of these labels
  repeated string tags = 29;

  // TODO
  // string start_datetime = 1;
  // string end_datetime = 2;
}

message VersionedGraft {
  // "1.2.3" - should adhere to semantic versioning.
  // See https://semver.org/
  string version = 1;
  string docstring = 5;
  string serialized_graft = 2;
  Typespec typespec = 4;

  // `parameters` are metadata about the Function's arguments.
  // If `typespec` describes a Function, there must be one `parameters` entry
  // for each positional argument, and its `name` must match the name of the corresponding
  // parameter in the `serialized_graft`.
  // Otherwise, `parameters` must be empty.
  repeated Parameter parameters = 11;
  string channel = 3;
  string client_version = 10;

  map<string, string> labels = 6; // mutable

  int64 created_timestamp = 7;
  int64 updated_timestamp = 8;
  int64 deprecated_timestamp = 9; // mutable

  // Template string for XYZ tiles like
  // "https://workflows.descarteslabs.com/v0-5/workflow/test@example.com:foo/1.0.1/{z}/{x}/{y}.png".
  // Compatible with Leaflet, OpenLayers, Python format strings.
  // Read-only: any value is ignored in `UpsertWorkflowRequest`.
  string url_template = 12;
  // Ditto for WMTS service endpoint.
  string wmts_url_template = 13;

  repeated VizOption viz_options = 14;
}

message WmtsUrlTemplateResponse {
  // Template string for WMTS service endpoint, like
  // "https://workflows.descarteslabs.com/v0-5/workflow?tile_matrix_set={TileMatrixSet}".
  // Compatible with Leaflet, OpenLayers, Python format strings.
  // Read-only: any value is ignored in `UpsertWorkflowRequest`.
  string wmts_url_template = 1;
}

message Empty {}
